src/test/debugger/appWorker.test.ts

```diff
@@ -0,0 +1,206 @@
+            const debugAdapterPort = 9090;
+
+            let sandboxedWorker: AppWorker.SandboxedAppWorker;
+            let downloadAppScriptStub = sinon.stub();
+            let postReplyFunction = sinon.stub();
+            let readFileStub = sinon.stub();
+
+            setup(function() {
+                const nodeFileSystemMock: any = {
+                    readFile: readFileStub
+                };
+
+                const scriptImporterMock: any = {
+                    downloadAppScript: downloadAppScriptStub
+                };
+
+                sandboxedWorker = new AppWorker.SandboxedAppWorker(sourcesStoragePath, debugAdapterPort, postReplyFunction, {
+                    nodeFileSystem: nodeFileSystemMock,
+                    scriptImporter: scriptImporterMock
+                });
+            });
+
+            teardown(function() {
+                // Reset everything
+                sandboxedWorker = null;
+
+                readFileStub = sinon.stub();
+                downloadAppScriptStub = sinon.stub();
+                postReplyFunction = sinon.stub();
+            });
+
+
+            test("should execute scripts correctly and be able to invoke the callback", function() {
+                const expectedMessageResult = { success: true };
+                const startScriptContents = `var testResponse = ${JSON.stringify(expectedMessageResult)}; postMessage(testResponse);`;
+
+                readFileStub.withArgs(startScriptFileName).returns(Q.resolve(startScriptContents));
+
+                return sandboxedWorker.start().then(() => {
+                    assert(postReplyFunction.calledWithExactly(expectedMessageResult));
+                });
+            });
+
+            test("should be able to import scripts", function() {
+                const scriptImportPath = "testScript.js";
+                const startScriptContents = `importScripts("${scriptImportPath}"); postMessage("postImport");`;
+                readFileStub.withArgs(startScriptFileName).returns(Q.resolve(startScriptContents));
+
+                const testScriptContents = "postMessage('inImport')";
+                const scriptImportDeferred = Q.defer<void>();
+                downloadAppScriptStub.withArgs(scriptImportPath, debugAdapterPort).returns(scriptImportDeferred.promise.then(() => {
+                    return {
+                        contents: testScriptContents,
+                        filepath: scriptImportPath
+                    };
+                }));
+
+                return sandboxedWorker.start().then(() => {
+                    // We have not yet finished importing the script, we should not have posted a response yet
+                    assert(postReplyFunction.notCalled, "postReplyFuncton called before scripts imported");
+                    scriptImportDeferred.resolve(void 0);
+                    return Q.delay(1);

```

Comment:
I'm not super happy about this, but I'm not sure how better to deal with these cases: We have some asynchronous operations that do not wait for a specified point in time and aren't chained into accessible promises, so to check that they happen I've just been waiting a minimal period of time to ensure that async operations have a chance to execute.

An alternative would be to explicitly wait on the callbacks that I expect, does that seem like a better approach? However in the case that they don't get called we end up with a timeout rather than a failed assertion.